### ISSUE TITLE
[Feature Request] Confirmation dialog when closing a modal window with changed content

### ISSUE THREAD

**Issue by redge76** [hargata/lubelog#444]
Hi,

Can you add a confirmation dialog when the escape key is pressed and data were added or changed in the modal window ?
Just lost some work in a note by pressing the escape key by mistake.
Something like the "Cofirm deletion" dialog when you click delete.

thanks

**Comment by hargata** [comment]
@redge76 thanks for submitting this, do you usually use the ESC key to close out the dialog or by clicking the "X" or "Cancel" buttons at the bottom of the dialog? I'm asking because I will be adding your feature in the next release but I'm contemplating if I should also disable closing the dialog via the ESC key.

**Comment by redge76** [comment]
I use ESC quite often while browsing rapidly through the notes. (it's easy to have the mouse cursor clicking on the list items sequencially and the other hand just hitting ESC. otherwise you have to click list->click cancel -> click list -> ...)
I never use X. I use cancel if I'm typing something and I want to abandon my entry. Then I consciously click on cancel.
So my main gripe is with ESC as it is quite easy to hit it accidentally or by "reflex". (Hey VIM users !! )
But the same dialog box for cancel and X would be great. (but only if you manage to make the confirmation appear if new data were entered)
In Gmail for example, you can use ESC but the email you are currently editing is saved as draft.
In https://trello.com/ if you hit ESC, the note is stored as "unsaved data" that you can recover later.

I guess it's a bit complicated to also save as draft in lubelog, but as "are you sure" dialog box would be great.

**Comment by redge76** [comment]
The confirmation dialog seems sufficient.

The CTRL+Z feature seems quite complicated and hard to use on mobile. KISS... Keep it simple .... :-P
(there is a lot of new features. So far it did not complicate the UI too much. But I like that LL is simple and I hope it will stay simple)

**Comment by hargata** [comment]
@redge76 the problem with the dialog feature is that it's very tricky to capture when the fields have changed across all browsers and all devices, I really wish it was as simple as binding the input event on each field and triggering a flag to let the modal know that the content has changed, but that's only possible if all of the fields are text fields, which isn't true since we have dropdowns, checkboxes / switches and file uploads in the modal as well. We have also tried serializing the form data and comparing it when the modal is loaded vs when it's closing, but that proved to be fairly expensive. Hence, we decided to move away from the dialog approach since we cannot guarantee that users' experience will be consistent across the board depending on which browser, browser version, and device they're using.

The CTRL+Z approach, while not perfect since it excludes mobile users, is extremely consistent across the board provided that the user has a keyboard. It doesn't actually have to check if the user has made any changes, but rather if the user pulled up a record and then close the modal without saving/moving/deleting the record, it caches the state of the modal including any unsaved changes and the user can bring the modal back up using CTRL+Z.

**Comment by redge76** [comment]
OK. This is how they solved this in trello. Maybe they had the same inconsistency between browsers.
If you hit ESC without saving, you can reopen the same card and a "unsaved changes" messages will appear with an edit button.
If you click edit you recover the last unsaved changes. If you start editing the changes are lost.
![image](https://github.com/hargata/lubelog/assets/1172508/3fac453d-4448-4474-b7b5-2a7efc4d1f17)
---> click edit
![image](https://github.com/hargata/lubelog/assets/1172508/5135e7d1-7523-4f03-8a64-cf972d919864)
display unsaved changes 
Now you can click save

**Comment by hargata** [comment]
@redge76 thanks, that does seem like a better approach for this. 
Here's what I've built so far:
When the user goes to edit/view an existing record, we're going to create a cached view of the record, and this cached view will contain any unsaved changes made by the user.
If the user closes the modal, whether intentionally or by accident, the cached view will persist until they either select a different record, switch tabs, refresh the page, or if the user saves, delete, or move the record.
If the user selects the same record as they have viewed previously, it will bring up the cached view including any unsaved changes.
When the cached view is displayed, the modal title will have a watermark saying "Cached"
If the user wants to display the latest version of the record, they can click on the "Cached" watermark and it will discard any unsaved changes they have and pull the latest version of the record from the database.

![cachedview](https://github.com/hargata/lubelog/assets/155338622/24ae7836-2d57-40a3-b431-77f9e2eabdd0)

I use the wording "Cached" instead of "Unsaved Changes" because the app will display the Cached version regardless of whether the user has made any changes to the record or not. It is basically saying, "hey, instead of retrieving a new copy of the same record you viewed previously from the database, here's a cached version"

I think this is about as good as it can get for now. I contemplated adding a "View Unsaved Changes" to highlight the fields that contains the differences or perhaps even making the cached view persist for more than just one record at a time, but that was too complex/expensive to run.
